src/pages/resultsView/pathwayMapper/ResultsViewPathwayMapper.tsx

```diff
@@ -0,0 +1,288 @@
+            this.activeToasts.length = 0;
+            toast.dismiss();
+            return undefined;
+        }
+
+        if (this.urlWrapperForAllGenes) {
+            if (
+                this.urlWrapperForAllGenes.hasSessionId &&
+                this.urlWrapperForAllGenes.remoteSessionData.isPending
+            ) {
+                return undefined;
+            }
+
+            const tId = toast(
+                'Alteration data of genes not listed in gene list might take a while to load!',
+                { autoClose: false, position: 'bottom-left' }
+            );
+            this.activeToasts.push(tId);
+
+            return initStore(this.props.appStore, this.urlWrapperForAllGenes);
+        }
+    }
+
+    @computed get urlWrapperForAllGenes(): ResultsViewURLWrapper | undefined {
+        let urlWrapper: ResultsViewURLWrapper | undefined;
+
+        if (
+            this.validNonQueryGenes.isComplete &&
+            this.validNonQueryGenes.result.length > 0
+        ) {
+            // fake the URL wrapper, we only need the query parameters with additional genes
+            const query: { [key: string]: string | undefined } = _.cloneDeep(
+                this.props.urlWrapper.query
+            );
+            query.gene_list = this.validNonQueryGenes.result.join(' ');
+
+            urlWrapper = { query } as ResultsViewURLWrapper;
+        }
+
+        return urlWrapper;
+    }
+
+    @computed get validGenes() {
+        if (this.validNonQueryGenes.isComplete) {
+            // Valid genes are accumulated.
+            this.validNonQueryGenes.result.forEach(gene => {
+                this.accumulatedValidGenes[gene] = true;
+            });
+        }
+        return this.accumulatedValidGenes;
+    }
+
+    // addGenomicData function is implemented in PathwayMapper component and overlays
+    // alteration data onto genes. Through this function callback, the function implemented
+    // in PathwayMapper is copied here.
+    @autobind
+    @action
+    private addGenomicDataHandler(
+        addGenomicData: (alterationData: ICBioData[]) => void
+    ) {
+        this.addGenomicData = addGenomicData;
+    }
+
+    // When pathway changes in PathwayMapper this callback gets called
+    @autobind
+    @action
+    private handlePathwayChange(pathwayGenes: string[]) {
+        // Pathway genes here are the genes that are in the pathway and valid whose alteration data is not calculated yet.
+        // Pathway genes does NOT always include all of the non-query genes
+        this.newGenesFromPathway = pathwayGenes;
+    }
+
+    @autobind
+    @action
+    private handleAddGenes(selectedGenes: string[]) {
+        // add new genes and go to oncoprint tab
+        const geneList = this.props.urlWrapper.query.gene_list;
+
+        this.props.urlWrapper.updateURL(
+            {
+                gene_list: `${geneList}\n${selectedGenes.join(' ')}`,
+            },
+            `results/${ResultsViewTab.ONCOPRINT}`
+        );
```

Comment:
can you pull the logic out from `AddCheckedGenes.tsx::onAddGenes` into a helper function and reuse it here?

Reply:
added a new `onAddGenes` helper function in `ResultsViewPageHelpers.ts`